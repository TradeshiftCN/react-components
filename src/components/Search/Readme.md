### Basic Usage

_Note: Normally onSearch callback will be called when press enter key and click clear btn_

```jsx
<Search
	placeholder="Search amongst all the things"
	onSearch={value => alert(`Your input is '${value}'`)}
/>
```

### Debounce

_Note: That will listen input/change event additionally and call onSearch after delay_

```jsx
initialState = { value: '' };
<div>
	<Search onSearch={value => setState({ value })} idleTime={500} />
	{state.value && (
		<p
			style={{
				marginTop: 16
			}}
		>
			Your input is '{state.value}'
		</p>
	)}
</div>;
```
