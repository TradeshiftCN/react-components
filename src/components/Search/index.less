@import '../../style/variables/index';

.ts-rc-search {
	position: relative;
	line-height: @ts-unit-double;
	i {
		position: absolute;
		top: 0;
		display: inline-block;
		width: @ts-unit-double;
		height: @ts-unit-double;
		line-height: inherit;
		vertical-align: middle;
		font-size: @ts-fontsize-icon;
	}
	input {
		line-height: @ts-unit-double - 2px !important;
		padding: 0 @ts-unit-double;
		border: 1px solid @ts-color-gray-light;
		width: 100%;
		display: block;
		border-radius: @ts-unit;
		&:hover,
		&:focus {
			box-shadow: 0 0 0 1px @ts-color-blue-light;
		}
		&:hover {
			border-color: @ts-color-blue-lighter;
		}
		&:focus {
			border-color: @ts-color-blue;
		}
	}
	&-search-icon {
		// FIXME: the baseline of this icon has a little offset
		padding-top: 1px;
		left: 0;
		font-weight: @ts-fontweight-semibold;
	}
	&-remove-icon {
		right: 0;
		color: @ts-color-slate-lightest;
		cursor: pointer;
	}
}
