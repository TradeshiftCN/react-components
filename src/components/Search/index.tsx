import React, { Component } from 'react';
import cx from 'classnames';
import _ from 'lodash';
import PropTypes from 'prop-types';

export type SearchProps = React.HTMLProps<HTMLInputElement> &
	React.HTMLAttributes<HTMLInputElement> & {
		value?: string;
		defaultValue?: string;
		onChange?(value: string): void;
		onSearch?(value: string): void;
		idleTime?: number;
		className?: string;
	};

type SearchState = {
	value: string;
	prevPropsValue: string;
};

const prefixCls = 'ts-rc-search';

class Search extends Component<SearchProps, SearchState> {
	static propTypes = {
		value: PropTypes.string,
		defaultValue: PropTypes.string,
		/** (value: string)=> void */
		onChange: PropTypes.func,
		/** (value: string)=> void */
		onSearch: PropTypes.func,
		/** call onSearch when triggered input event and idleTime as delay */
		idleTime: PropTypes.number,
		className: PropTypes.string
	};
	public state: SearchState;
	private inputRef?: HTMLInputElement | null;
	private debounceSubmit?: ReturnType<typeof _.debounce>;
	constructor(props: SearchProps) {
		super(props);
		const value = !_.isNil(props.value) ? props.value : props.defaultValue || '';
		this.state = {
			value,
			prevPropsValue: value
		};
		this.updateDebounceSubmit(props.onSearch, props.idleTime);
	}
	render() {
		const { className } = this.props;
		const inputProps = _.omit(this.props, [
			'value',
			'defaultValue',
			'onChange',
			'onSearch',
			'idleTime',
			'className'
		]);
		const { value } = this.state;

		return (
			<div className={cx(prefixCls, className)}>
				<i className={`ts-icon-search ${prefixCls}-search-icon`}></i>
				<input
					type="text"
					value={value}
					onChange={this.handleChange}
					onKeyDown={this.handleKeyDown}
					ref={ref => (this.inputRef = ref)}
					{...inputProps}
				/>
				{!!value && (
					<i
						className={`ts-icon-remove ${prefixCls}-remove-icon`}
						onClick={this.handleClickRemoveBtn}
					></i>
				)}
			</div>
		);
	}

	// FIXME: support react v15
	public componentWillReceiveProps(nextProps: SearchProps) {
		const { value = '' } = nextProps;
		if (value !== this.state.prevPropsValue) {
			this.setState({
				value,
				prevPropsValue: value
			});
		}

		if (this.props.onSearch !== nextProps.onSearch || this.props.idleTime !== nextProps.idleTime) {
			this.updateDebounceSubmit(nextProps.onSearch, nextProps.idleTime);
		}
	}

	private updateValue(value: string) {
		const { onChange } = this.props;

		this.setState({
			value
		});

		if (onChange) {
			onChange(value);
		}

		if (this.debounceSubmit) {
			this.debounceSubmit(value);
		}
	}

	private manualSearch(value: string) {
		if (this.props.onSearch) {
			this.props.onSearch(value);
			if (this.debounceSubmit) {
				this.debounceSubmit.cancel();
			}
		}
	}

	private handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.updateValue(e.currentTarget.value);
	};

	private handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
		if (this.props.onKeyDown) {
			this.props.onKeyDown(e);
		}
		if (e.keyCode === 13) {
			e.preventDefault();
			this.manualSearch(e.currentTarget.value);
		}
	};

	private handleClickRemoveBtn = () => {
		this.updateValue('');
		this.focusInput();
	};

	private focusInput() {
		if (this.inputRef) {
			this.inputRef.focus();
		}
	}

	private updateDebounceSubmit(onSearch?: (value: string) => void, idleTime?: number) {
		if (onSearch && idleTime) {
			this.debounceSubmit = _.debounce((v: string) => {
				onSearch(v);
			}, idleTime);
		} else {
			this.debounceSubmit = undefined;
		}
	}
}

export default Search;
