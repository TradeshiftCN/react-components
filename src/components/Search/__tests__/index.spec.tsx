import React from 'react';
import { shallow } from 'enzyme';
import Search from '..';

describe('Search', () => {
	const wait = async (time: number) => new Promise(resolve => setTimeout(resolve, time));
	it('basic', () => {
		const wrapper = shallow(<Search />);
		expect(wrapper).toMatchSnapshot();
	});

	it('update value', () => {
		const onChange = jest.fn();
		const wrapper = shallow(<Search onChange={onChange} />);
		const input = wrapper.find('input');
		input.simulate('change', {
			currentTarget: {
				value: 'test'
			}
		});

		expect(wrapper.state('value')).toBe('test');
		expect(onChange.mock.calls[0][0]).toBe('test');
		onChange.mockClear();

		const clearBtn = wrapper.find('.ts-rc-search-remove-icon');
		expect(clearBtn.exists()).toBe(true);
		clearBtn.simulate('click');
		expect(wrapper.state('value')).toBe('');
		expect(onChange.mock.calls[0][0]).toBe('');
	});

	it('render props value', () => {
		const wrapper = shallow(<Search value="value" />);
		expect(wrapper.find('input').props().value).toBe('value');
	});

	it('defaultValue', () => {
		let wrapper;

		wrapper = shallow(<Search defaultValue="default" />);
		expect(wrapper.find('input').props().value).toBe('default');
		wrapper = shallow(<Search value="current" defaultValue="default" />);
		expect(wrapper.find('input').props().value).toBe('current');
	});

	it('search', () => {
		const onSearch = jest.fn();
		const wrapper = shallow(<Search onSearch={onSearch} value="test" />);
		const input = wrapper.find('input');
		input.simulate('keyDown', {
			currentTarget: {
				value: 'value'
			},
			keyCode: 13,
			preventDefault() {}
		});

		expect(onSearch.mock.calls[0][0]).toBe('value');
		onSearch.mockClear();
		wrapper.find('.ts-rc-search-remove-icon').simulate('click');
		expect(wrapper.state('value')).toBe('');
		expect(onSearch).not.toHaveBeenCalled();
	});

	it('debounce', async () => {
		const onSearch = jest.fn();
		const wrapper = shallow(<Search onSearch={onSearch} idleTime={100} />);
		const input = wrapper.find('input');
		input.simulate('change', {
			currentTarget: {
				value: 'test'
			}
		});
		expect(onSearch).not.toHaveBeenCalled();
		await wait(100);
		expect(onSearch.mock.calls[0][0]).toBe('test');
	});
});
