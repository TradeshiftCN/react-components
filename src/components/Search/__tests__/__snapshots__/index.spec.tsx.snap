// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search basic 1`] = `
<div
  className="ts-rc-search"
>
  <i
    className="ts-icon-search ts-rc-search-search-icon"
  />
  <input
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
</div>
`;
